### PR TITLE
IBX-1489: Provided integration test for Installers for PHP8

### PR DESCRIPTION
**Pull request by Steveb-p**
| Question                                  | Answer
| ---------------------------------------- | ------------------
| **JIRA issue**                          | [IBX-1489](https://issues.ibexa.co/browse/IBX-1489)
| **Type**                                   | improvement
| **Target Ibexa version** | `v3.3`
| **BC breaks**                          | no

Provides integration tests & fixes for PHP8 compatibility.

# Notable changes:

## Disabling transactions during schema import / install.

Due to a change in PHP's MySQL driver https://github.com/php/php-src/commit/990bb34 for PHP8 it becomes necessary to handle a case where `Connection::commit()` throws an exception that there is no transaction (which happens after DDL SQL commands).

> DDL statements on mysql result in an implicit commit (see [mysql documentation](https://dev.mysql.com/doc/refman/8.0/en/implicit-commit.html)). This causes the explicit commit in transactional mode to fail. But only starting with php 8.0 an error is raised (probably due to php/php-src@990bb34).

## Removal of PHP8 deprecation warnings

This includes:
 * Reordering of some test methods, so that optional argument are at the end of the list.
 * Converting values to `string`, where other data types are possible. Especially `null` or `int`.

## Issues that will require addressing in the near future
 * One test fails specifically on PHP 8.1 due to change in `DateTime` interval behavior. It is therefore disabled on this platform.
 * MySQL 8.0 tests have issues with collation conversion (which are **probably** silently ignored on earlier versions).

#### Checklist:
- [x] Provided PR description.
- [ ] Tested the solution manually.
- [x] Provided automated test coverage.
- [x] Checked that target branch is set correctly (master for features, the oldest supported for bugs).
- [x] Ran PHP CS Fixer for new PHP code (use `$ composer fix-cs`).
- [x] Asked for a review (ping `@ezsystems/engineering-team`).

### CODE REVIEW

File: .github/workflows/ci.yaml [hunk 1]
```diff
@@ -21,6 +21,9 @@ jobs:
                     - '7.3'
                     - '7.4'
                 composer_options: [ "" ]
+                include:
+                    -   php: '8.0'
```
Comment (konradoboza) [inline .github/workflows/ci.yaml:25]: We might consider switching to 8.1 since it was released today.

Comment (Steveb-p) [reply .github/workflows/ci.yaml:25]: On kernel/core we can safely add both, because open source projects have unlimited GH Action minutes.

File: .github/workflows/ci.yaml [hunk 2]
```diff
@@ -17,10 +17,17 @@ jobs:
             fail-fast: false
             matrix:
                 experimental: [ false ]
+                code_style_check: [ true ]
```
Comment (Steveb-p) [inline .github/workflows/ci.yaml]:
```suggestion
```

File: .github/workflows/ci.yaml [hunk 3]
```diff
@@ -17,10 +17,17 @@ jobs:
...
                 php:
                     - '7.3'
                     - '7.4'
                 composer_options: [ "" ]
+                include:
+                    -   php: '8.0'
+                        composer_options: "--ignore-platform-req php"
+                    -   php: '8.1'
+                        composer_options: "--ignore-platform-req php"
+                        code_style_check: false
```
Comment (Steveb-p) [inline .github/workflows/ci.yaml]:
```suggestion
                        skip_code_style: true
```

File: .github/workflows/ci.yaml [hunk 4]
```diff
@@ -45,6 +52,7 @@ jobs:
             #   run: composer run-script phpstan
 
             - name: Run code style check
+              if: matrix.code_style_check
```
Comment (Steveb-p) [inline .github/workflows/ci.yaml]: This change will make it so we don't need to explicitly declare in matrix that we want code style to run. It will run automatically unless we explicitly say otherwise.
```suggestion
              if: !matrix.skip_code_style
```

File: eZ/Publish/Core/FieldType/Media/Type.php [hunk 5]
```diff
@@ -246,21 +246,11 @@ public function fromPersistenceValue(FieldValue $fieldValue)
...
-            ? $fieldValue->externalData['height']
-            : 0);
-        $result->width = (isset($fieldValue->externalData['width'])
-            ? $fieldValue->externalData['width']
-            : 0);
+        $result->hasController = ($fieldValue->externalData['hasController'] ?? false);
+        $result->autoplay = ($fieldValue->externalData['autoplay'] ?? false);
+        $result->loop = ($fieldValue->externalData['loop'] ?? false);
+        $result->height = ($fieldValue->externalData['height'] ?? 0);
+        $result->width = ($fieldValue->externalData['width'] ?? 0);
```
Comment (Steveb-p) [inline eZ/Publish/Core/FieldType/Media/Type.php]:
```suggestion
        $result->hasController = $fieldValue->externalData['hasController'] ?? false;
        $result->autoplay = $fieldValue->externalData['autoplay'] ?? false;
        $result->loop = $fieldValue->externalData['loop'] ?? false;
        $result->height = $fieldValue->externalData['height'] ?? 0;
        $result->width = $fieldValue->externalData['width'] ?? 0;
```

File: .github/workflows/ci.yaml [hunk 6]
```diff
@@ -55,7 +63,7 @@ jobs:
 
     integration-tests-postgres:
         name: PostgreSQL integration tests
-        needs: tests
+        # needs: tests
```
Comment (Steveb-p) [inline .github/workflows/ci.yaml]: We should remember to revert this one (although technically it's OSS, so we could run integration tests against databases immediately - github pays :dollar: )

File: eZ/Publish/Core/MVC/Symfony/FieldType/Tests/User/ParameterProviderTest.php [hunk 7]
```diff
@@ -44,6 +44,9 @@ protected function setUp(): void
         $this->parameterProvider = new ParameterProvider($this->userService);
     }
 
+    /**
+     * @requires PHP < 8.1
```
Comment (konradoboza) [inline eZ/Publish/Core/MVC/Symfony/FieldType/Tests/User/ParameterProviderTest.php:48]: The reason for that was most likely some regression coming from PHP 8.1, ref: https://github.com/ezsystems/ezplatform-kernel/runs/4354252098?check_suite_focus=true#step:7:156.

File: .github/workflows/ci.yaml [hunk 8]
```diff
@@ -72,16 +79,19 @@ jobs:
...
             fail-fast: false
             matrix:
-                experimental: [ false ]
                 php:
                     - '7.3'
                     - '7.4'
                 composer_options: [ "" ]
+                include:
+                    -   php: '8.0'
+                        composer_options: "--ignore-platform-req php"
```
Comment (alongosz) [inline .github/workflows/ci.yaml:92]: If you have:
```json
"php": "^7.3 || ^8.0",
```
Do you still need `--ignore-platform-req php`?

**Comment by konradoboza** [comment]
The failing test case seems to be random, given it passes on all other PHP versions. Perhaps tests order comes into play.

File: eZ/Bundle/PlatformInstallerBundle/src/Installer/CoreInstaller.php [hunk 9]
```diff
@@ -67,16 +66,9 @@ public function importSchema()
...
-                $this->db->exec($query);
-                $progressBar->advance(1);
-            }
-            $this->db->commit();
-        } catch (DBALException $e) {
-            $this->db->rollBack();
-            throw $e;
+        foreach ($queries as $query) {
+            $this->db->exec($query);
+            $progressBar->advance(1);
```
Comment (alongosz) [inline eZ/Bundle/PlatformInstallerBundle/src/Installer/CoreInstaller.php:71]: Why do you need to get rid of this? If there's no transaction, a transaction is per commit which means that in case of any failure, partial installation will be persisted.

File: eZ/Publish/Core/Persistence/Tests/Legacy/FieldValue/Converter/ImageConverterTest.php [hunk 10]
```diff
@@ -48,7 +48,7 @@ public function testToStorageValue(FieldValue $fieldValue, string $expectedXml):
         ClockMock::withClockMock(true);
 
         $time = ClockMock::time();
-        $expectedXml = str_replace('{timestampToReplace}', $time, $expectedXml);
+        $expectedXml = str_replace('{timestampToReplace}', (string) $time, $expectedXml);
```
Comment (alongosz) [inline eZ/Publish/Core/Persistence/Tests/Legacy/FieldValue/Converter/ImageConverterTest.php]: inconsistent CS, so far you had no space between a cast and a variable.

File: eZ/Bundle/EzPublishCoreBundle/Imagine/AliasGenerator.php [hunk 11]
```diff
@@ -94,7 +94,7 @@ public function getVariation(Field $field, VersionInfo $versionInfo, $variationN
             try {
                 $originalBinary = $this->dataLoader->find($originalPath);
             } catch (NotLoadableException $e) {
-                throw new SourceImageNotFoundException($originalPath, 0, $e);
+                throw new SourceImageNotFoundException((string)$originalPath, 0, $e);
```
Comment (alongosz) [inline eZ/Bundle/EzPublishCoreBundle/Imagine/AliasGenerator.php:97]: Why is this necessary and how it's related to PHP 8, since `SourceImageNotFoundException` is our own class?

File: eZ/Publish/Core/FieldType/Tests/AuthorTest.php [hunk 12]
```diff
@@ -561,8 +561,8 @@ public function provideDataForGetName(): array
         );
 
         return [
-            [$this->getEmptyValueExpectation(), [], 'en_GB', ''],
-            [$authorList, [], 'en_GB', 'Boba Fett'],
+            [$this->getEmptyValueExpectation(), '', [], 'en_GB'],
+            [$authorList, 'Boba Fett', [], 'en_GB'],
```
Comment (alongosz) [inline eZ/Publish/Core/FieldType/Tests/AuthorTest.php:565]: Why do we change the order of arguments generated by the provider and how is it related to PHP 8?

**Review by alongosz — CHANGES_REQUESTED** [review]
It's a bit unclear to me why some changes were made and how they're related to what the PR is supposed to be fixing. Unfortunately the PR description is rather empty.

File: .github/workflows/ci.yaml (hunk 8 above)
Comment (Steveb-p) [reply .github/workflows/ci.yaml:92]: Other dependencies prevent us from not using this option.

```
composer update
Loading composer repositories with package information
Updating dependencies
Your requirements could not be resolved to an installable set of packages.

  Problem 1
    - Root composer.json requires php ^7.3 but your php version (8.0.11) does not satisfy that requirement.
  Problem 2
    - Root composer.json requires ibexa/ci-scripts ^0.1@dev -> satisfiable by ibexa/ci-scripts[0.1.x-dev].
    - ibexa/ci-scripts 0.1.x-dev requires php ^7.3 -> your php version (8.0.11) does not satisfy that requirement.
  Problem 3
    - ezsystems/doctrine-dbal-schema[v1.0.0-beta1, ..., 1.0.x-dev] require php ^7.3 -> your php version (8.0.11) does not satisfy that requirement.
    - Root composer.json requires ezsystems/doctrine-dbal-schema ^1.0@dev -> satisfiable by ezsystems/doctrine-dbal-schema[v1.0.0-beta1, ..., 1.0.x-dev].
```

File: eZ/Bundle/PlatformInstallerBundle/src/Installer/CoreInstaller.php (hunk 9 above)
Comment (Steveb-p) [reply eZ/Bundle/PlatformInstallerBundle/src/Installer/CoreInstaller.php:71]: MySQL driver received a change in PHP 8.0 that makes attempting to commit a transaction when there isn't any an error. Because DDL operations do not allow transactions in MySQL this means that transaction is implicitly committed. This results in an error.

A similar solution was made for doctrine/migrations, where it was suggested that mysql Migrations should explicitly be made outside of transaction. See https://github.com/doctrine/migrations/issues/1104.

Note that technically only Postgres supports transactions in DDL operations, which means that in reality on MySQL those were in fact partial installations.

Another option is acquiring internal `PDO` object from Doctrine DBAL's `Connection` (via `getWrappedConnection()`) and using `inTransaction()` method which asks the database if there is a transaction in progress (which is also one of the suggested solutions to the issue, and which is cached for subsequent calls if there were no SQL commands).

File: eZ/Bundle/EzPublishCoreBundle/Imagine/AliasGenerator.php (hunk 11 above)
Comment (Steveb-p) [reply eZ/Bundle/EzPublishCoreBundle/Imagine/AliasGenerator.php:97]: I'm almost certain that this intends to fix an issue when `null` value is in `$originalPath`. Possibly some tests do not configure a mock I guess?

File: eZ/Bundle/PlatformInstallerBundle/src/Installer/CoreInstaller.php (hunk 9 above)
Comment (konradoboza) [reply eZ/Bundle/PlatformInstallerBundle/src/Installer/CoreInstaller.php:71]: As of PHP 8, `PDO::commit()` and `rollBack()` will throw a `There is no active transaction` PDOException if the transaction had been autocommitted, Previously they would just fail silently.

File: eZ/Bundle/EzPublishCoreBundle/Imagine/AliasGenerator.php (hunk 11 above)
Comment (konradoboza) [reply eZ/Bundle/EzPublishCoreBundle/Imagine/AliasGenerator.php:97]: This fix aims to suppress deprecation warnings like: `PHP Deprecated:  Exception::__construct(): Passing null to parameter #1 ($message) of type string is deprecated`.

File: eZ/Publish/Core/FieldType/Tests/AuthorTest.php (hunk 12 above)
Comment (konradoboza) [reply eZ/Publish/Core/FieldType/Tests/AuthorTest.php:565]: Another deprecation warning: `PHP Deprecated:  Optional parameter [xxx] declared before required parameter [yyy] is implicitly treated as a required parameter`, ref: https://www.php.net/manual/en/migration80.deprecated.php#migration80.deprecated.core.

Comment (Steveb-p) [reply eZ/Publish/Core/FieldType/Tests/AuthorTest.php:565]: From what I see the changes are limited to test classes only. If it were otherwise, I would say see we should make the arguments mandatory instead and remove the default value (as it was impossible to use the optional parameter anyway - it always had to be specified).

That is, unless those test classes were used in other libraries, which I know might have been the case sometimes?

Comment (konradoboza) [reply eZ/Publish/Core/FieldType/Tests/AuthorTest.php:565]: >From what I see the changes are limited to test classes only

This is the main reason why I decided to just switch arguments order. I am not sure if we should care about providing any BC for tests anyway.

File: eZ/Bundle/EzPublishCoreBundle/Imagine/VariationPurger/LegacyStorageImageFileList.php [hunk 13]
```diff
@@ -96,7 +97,7 @@ private function fetchRow(): void
         ++$this->cursor;
         $imageId = $this->rowReader->getRow();
 
-        if (substr($imageId, 0, strlen($prefix)) === $prefix) {
+        if (strpos((string)$imageId, $prefix) === 0) {
```
Comment (Steveb-p) [inline eZ/Bundle/EzPublishCoreBundle/Imagine/VariationPurger/LegacyStorageImageFileList.php]: As far as I know the previous variant would be _slightly_ more efficient (but that's micro). What was the reason for change?

Comment (konradoboza) [reply eZ/Bundle/EzPublishCoreBundle/Imagine/VariationPurger/LegacyStorageImageFileList.php]: This was one of the SonarLint plugin suggestions. To me, readability benefits here, but this isn't something I would have any strong opinion on.

Comment (Steveb-p) [reply eZ/Bundle/EzPublishCoreBundle/Imagine/VariationPurger/LegacyStorageImageFileList.php]: Symfony polyfill for PHP 7.4 proposes the following implementation of `str_starts_with`:
```php
public static function str_starts_with(string $haystack, string $needle): bool
{
    return 0 === strncmp($haystack, $needle, strlen($needle));
}
```
Benefits include:
 * binary-safe comparison
 * only beginning of the string is compared, which is not the case for `strpos` (which looks up through the whole string)

File: .github/workflows/ci.yaml [hunk 14]
```diff
@@ -108,3 +118,62 @@ jobs:
                     DATABASE_URL: "pgsql://postgres:postgres@localhost:${{ job.services.postgres.ports[5432] }}/testdb?server_version=10"
                     # Required by old repository tests
                     DATABASE: "pgsql://postgres:postgres@localhost:${{ job.services.postgres.ports[5432] }}/testdb"
+
+    integration-tests-mysql:
+        name: MySQL integration tests
+        needs: tests
+        services:
+            mysql:
+                image: mysql:5.7
```
Comment (Steveb-p) [inline .github/workflows/ci.yaml:127]: Note: A follow up draft PR with `mysql:8.0` is necessary to show that there are tests that will fail on MySQL 8.

File: eZ/Bundle/PlatformInstallerBundle/src/Installer/CoreInstaller.php (hunk 9 above)
Comment (alongosz) [reply eZ/Bundle/PlatformInstallerBundle/src/Installer/CoreInstaller.php:71]: Fair enough, can't do much about that, +1

File: eZ/Bundle/EzPublishCoreBundle/Imagine/AliasGenerator.php (hunk 11 above)
Comment (alongosz) [reply eZ/Bundle/EzPublishCoreBundle/Imagine/AliasGenerator.php:97]: Thx for the explanation! +1

File: eZ/Publish/Core/FieldType/Tests/AuthorTest.php (hunk 12 above)
Comment (alongosz) [reply eZ/Publish/Core/FieldType/Tests/AuthorTest.php:565]: +1

**Review by alongosz — APPROVED** [review]
+1 but Steve has some valid remarks: https://github.com/ezsystems/ezplatform-kernel/pull/264#pullrequestreview-818744103 ;)

**Review by konradoboza — APPROVED** [review]

**Review by micszo — APPROVED** [review]
Tested with sanities on Ibexa Commerce 3.3.12-dev with PHP8.1.

QA Approved.
